File: src/ghsingo/internal/audio/reverb.go
package audio

// Reverb implements a simple Schroeder reverb using four parallel comb filters
// followed by two all-pass filters. It gives event sounds an underwater / large
// space quality without external DSP libraries.
//
// Usage:
//
//	r := NewReverb(sampleRate)
//	wetSample := r.Process(drySample)
type Reverb struct {
	combs    [4]combFilter
	allpasses [2]allpassFilter
	wet      float32 // wet mix level (0 = dry only, 1 = wet only)
}

// combFilter is a simple feedback comb filter: delay + feedback gain.
type combFilter struct {
	buf  []float32
	pos  int
	gain float32
}

// allpassFilter is a Schroeder all-pass section.
type allpassFilter struct {
	buf  []float32
	pos  int
	gain float32
}

// NewReverb creates a Reverb for the given sample rate.
// Wet mix defaults to 0.25 (25% reverb).
func NewReverb(sampleRate int) *Reverb {
	// Comb delays (ms) chosen as mutually prime to avoid metallic resonance:
	// 37ms, 41ms, 43ms, 47ms
	combDelaysMs := [4]float64{37, 41, 43, 47}
	combGains := [4]float32{0.80, 0.78, 0.76, 0.74}

	// All-pass delays: 5ms, 1.7ms
	apDelaysMs := [2]float64{5.0, 1.7}
	apGains := [2]float32{0.5, 0.5}

	r := &Reverb{wet: 0.25}

	for i := 0; i < 4; i++ {
		n := int(combDelaysMs[i] * float64(sampleRate) / 1000.0)
		r.combs[i] = combFilter{
			buf:  make([]float32, n),
			gain: combGains[i],
		}
	}
	for i := 0; i < 2; i++ {
		n := int(apDelaysMs[i] * float64(sampleRate) / 1000.0)
		if n < 1 {
			n = 1
		}
		r.allpasses[i] = allpassFilter{
			buf:  make([]float32, n),
			gain: apGains[i],
		}
	}
	return r
}

// SetWet sets the wet/dry mix (0.0 = fully dry, 1.0 = fully wet).
func (r *Reverb) SetWet(wet float32) {
	r.wet = wet
}

// Process applies one sample through the reverb and returns the wet+dry mix.
func (r *Reverb) Process(x float32) float32 {
	// Run four parallel comb filters and sum their output.
	var combSum float32
	for i := range r.combs {
		combSum += r.combs[i].process(x)
	}
	combSum *= 0.25 // normalise

	// Series all-pass filters.
	ap := r.allpasses[0].process(combSum)
	ap = r.allpasses[1].process(ap)

	return x*(1-r.wet) + ap*r.wet
}

// process runs one sample through a comb filter.
func (c *combFilter) process(x float32) float32 {
	out := c.buf[c.pos]
	c.buf[c.pos] = x + out*c.gain
	c.pos = (c.pos + 1) % len(c.buf)
	return out
}

// process runs one sample through an all-pass filter.
func (a *allpassFilter) process(x float32) float32 {
	delayed := a.buf[a.pos]
	out := -x + delayed
	a.buf[a.pos] = x + delayed*a.gain
	a.pos = (a.pos + 1) % len(a.buf)
	return out
}

File: src/rsghsing/src/audio/reverb.rs
//! Schroeder reverb: 4 parallel combs -> 2 series all-passes.
//! Ported from `internal/audio/reverb.go`.

pub struct Reverb {
    combs: [CombFilter; 4],
    allpasses: [AllpassFilter; 2],
    wet: f32,
}

struct CombFilter {
    buf: Vec<f32>,
    pos: usize,
    gain: f32,
}

struct AllpassFilter {
    buf: Vec<f32>,
    pos: usize,
    gain: f32,
}

impl CombFilter {
    fn process(&mut self, x: f32) -> f32 {
        let out = self.buf[self.pos];
        self.buf[self.pos] = x + out * self.gain;
        self.pos = (self.pos + 1) % self.buf.len();
        out
    }
}

impl AllpassFilter {
    fn process(&mut self, x: f32) -> f32 {
        let delayed = self.buf[self.pos];
        let out = -x + delayed;
        self.buf[self.pos] = x + delayed * self.gain;
        self.pos = (self.pos + 1) % self.buf.len();
        out
    }
}

impl Reverb {
    pub fn new(sample_rate: i32) -> Self {
        // Mutually-prime comb delays (ms) to avoid metallic resonance.
        let comb_delays_ms = [37.0f64, 41.0, 43.0, 47.0];
        let comb_gains = [0.80f32, 0.78, 0.76, 0.74];
        let ap_delays_ms = [5.0f64, 1.7];
        let ap_gains = [0.5f32, 0.5];

        let mut combs: [CombFilter; 4] = [
            CombFilter {
                buf: Vec::new(),
                pos: 0,
                gain: 0.0,
            },
            CombFilter {
                buf: Vec::new(),
                pos: 0,
                gain: 0.0,
            },
            CombFilter {
                buf: Vec::new(),
                pos: 0,
                gain: 0.0,
            },
            CombFilter {
                buf: Vec::new(),
                pos: 0,
                gain: 0.0,
            },
        ];
        for i in 0..4 {
            let n = (comb_delays_ms[i] * f64::from(sample_rate) / 1000.0) as usize;
            combs[i] = CombFilter {
                buf: vec![0.0; n.max(1)],
                pos: 0,
                gain: comb_gains[i],
            };
        }

        let mut allpasses: [AllpassFilter; 2] = [
            AllpassFilter {
                buf: Vec::new(),
                pos: 0,
                gain: 0.0,
            },
            AllpassFilter {
                buf: Vec::new(),
                pos: 0,
                gain: 0.0,
            },
        ];
        for i in 0..2 {
            let n = ((ap_delays_ms[i] * f64::from(sample_rate) / 1000.0) as usize).max(1);
            allpasses[i] = AllpassFilter {
                buf: vec![0.0; n],
                pos: 0,
                gain: ap_gains[i],
            };
        }

        Reverb {
            combs,
            allpasses,
            wet: 0.25,
        }
    }

    pub fn set_wet(&mut self, wet: f32) {
        self.wet = wet;
    }

    pub fn process(&mut self, x: f32) -> f32 {
        let mut comb_sum = 0.0f32;
        for c in &mut self.combs {
            comb_sum += c.process(x);
        }
        comb_sum *= 0.25;
        let mut ap = self.allpasses[0].process(comb_sum);
        ap = self.allpasses[1].process(ap);
        x * (1.0 - self.wet) + ap * self.wet
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Acceptance 3: the reverb must be stable for a 10-minute render —
    /// no NaN, no unbounded growth (all comb gains are < 1, so the loop is
    /// contractive; this pins that property).
    #[test]
    fn ten_minutes_of_signal_stays_finite_and_bounded() {
        let sample_rate = 44_100;
        let mut r = Reverb::new(sample_rate);
        r.set_wet(0.45);
        let mut peak = 0.0f32;
        for i in 0..(sample_rate * 600) {
            // Worst-case-ish drive: full-scale square-ish input.
            let x = if i % 2 == 0 { 1.0 } else { -1.0 };
            let y = r.process(x);
            assert!(y.is_finite(), "non-finite at sample {i}");
            peak = peak.max(y.abs());
        }
        assert!(peak < 4.0, "reverb diverged: peak {peak}");
    }

    #[test]
    fn wet_zero_returns_dry() {
        let mut r = Reverb::new(44_100);
        r.set_wet(0.0);
        assert_eq!(r.process(0.5), 0.5);
    }
}
